src/leo_ssec_base.erl

```diff
@@ -0,0 +1,354 @@
+%%
+-spec(stream_decrypt_data(UserKey, Data, IVec, AlgoMetaData) ->
+        {State, CipherData} when UserKey::{key, crypto:io_data()} | State,
+                                  Data::crypto:io_data(),
+                                  IVec::binary(),
+                                  AlgoMetaData::algo_metadata(),
+                                  State::{state, crypto:opaque()},
+                                  CipherData::binary()).
+stream_decrypt_data({key, UserKey}, Data, IVec, AlgoMetaData) ->
+    {Algo, _} = AlgoMetaData,
+    stream_decrypt_data({state, crypto:stream_init(Algo, UserKey, IVec)}, Data).
+
+%% @doc verify that block data is encrypted correctly
+%%
+-spec(verify_block_encryption(Key, Msg, AlgoMetaData) ->
+        {Status, EncryptedMsg} when Key::crypto:block_key(),
+                                    Msg::crypto:io_data(),
+                                    AlgoMetaData::algo_metadata(),
+                                    Status::boolean(),
+                                    EncryptedMsg::binary()).
+verify_block_encryption(Key, Msg, AlgoMetaData) ->
+    EncryptedMsg = block_encrypt_data(Key, Msg, AlgoMetaData),
+    {Msg =:= block_decrypt_data(Key, EncryptedMsg, AlgoMetaData),
+         EncryptedMsg}.
+
+-spec(verify_stream_encryption(_Key, _Msg, _AlgoMetaData) ->
+        {Status, EncryptedMsg} when _Key::crypto:stream_key(),
+                                    _Msg::crypto:io_data(),
+                                    _AlgoMetaData::algo_metadata(),
+                                    Status::boolean(),
+                                    EncryptedMsg::binary()).
+%% @doc verify that stream data is encrypted correctly
+verify_stream_encryption(_Key, _Msg, _AlgoMetaData) ->
+    {false, "@TODO not implemented yet"}.
+
+%% @doc verifies that the user demanded the correct algorithm
+%%      only AES256 allowed
+-spec(verify_ssec_algorithm(Algorithm) ->
+        {Status, ValidAlgorithms} when Algorithm::string(),
+                                       Status::boolean(),
+                                       ValidAlgorithms::[string()]).
+verify_ssec_algorithm(Algorithm) ->
+    ValidAlgorithms = ["AES256"],
+    {lists:member(Algorithm, ValidAlgorithms), ValidAlgorithms}.
+
+%% @doc verifies that the user provided key matches the hash
+%%      Key and hash are Base64 encoded
+-spec(verify_ssec_key(ASCIIKey, Checksum) ->
+        {Status, ErrorDescription} when ASCIIKey::string(),
+                                        Checksum::{Type, ASCIIHash},
+                                        Type::md5,
+                                        ASCIIHash::string(),
+                                        Status::boolean(),
+                                        ErrorDescription::string()).
+verify_ssec_key(ASCIIKey, Checksum) ->
+    Key = base64:decode(ASCIIKey),
+    {HashType, ASCIIHash} = Checksum,
+    HashValue = base64:decode(ASCIIHash),
+    if
+        size(Key) /= 256/4 ->
```

Comment:
Will correct this. I wrote a wrong test case. Supplied data to ```base:encode``` in Base 16, instead of binary. Inflated the input size. Should have thought twice before moving from 8 to 4.